playwright/src/main/java/com/microsoft/playwright/junit/impl/PlaywrightExtension.java

```diff
@@ -4,19 +4,51 @@
 import com.microsoft.playwright.junit.Options;
 import org.junit.jupiter.api.extension.*;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
 import static com.microsoft.playwright.junit.impl.ExtensionUtils.isParameterSupported;
 import static com.microsoft.playwright.junit.impl.ExtensionUtils.setTestIdAttribute;
 
-public class PlaywrightExtension implements ParameterResolver, AfterAllCallback {
-  private static final ThreadLocal<Playwright> threadLocalPlaywright = new ThreadLocal<>();
+public class PlaywrightExtension implements ParameterResolver, BeforeAllCallback, ExtensionContext.Store.CloseableResource {
+  private static final ThreadLocal<Playwright> threadLocalPlaywright;
+  private static final List<Playwright> playwrightList;
+  private final static Lock beforeLock;
+  private static boolean isTestRunStarted;
+
+  static {
+    isTestRunStarted = false;
+    beforeLock = new ReentrantLock();
+    threadLocalPlaywright = new ThreadLocal<>();
+    playwrightList = Collections.synchronizedList(new ArrayList<>());
+  }
 
+  // Before a Test class starts, we register a closeable resource
+  // We use a lock + boolean to ensure this only gets run once for the entire test run
   @Override
-  public void afterAll(ExtensionContext extensionContext) {
-    Playwright playwright = threadLocalPlaywright.get();
-    threadLocalPlaywright.remove();
-    if (playwright != null) {
+  public void beforeAll(final ExtensionContext context) {
+    try {
+      beforeLock.lock();
+      if (!isTestRunStarted) {
+        isTestRunStarted = true;
+        context.getRoot().getStore(ExtensionContext.Namespace.GLOBAL).put(context.getUniqueId(), this);
```

Comment:
Let's use Playwright specific namespace here to avoid possible clashes? Something like `ExtensionContext.Namespace.create(PlaywrightException.class)` should do.

Reply:
I'll need to look into this some more but I believe that closeable resources need to be registered in the global namespace. I don't think there will be conflicts with anything else because we are using this extension's context's unique id as the key.

But I'll play around with it to see if we can register a closeable resource in a named namespace.

Reply:
@yury-s verified that closeable resource can be registered in a named namespace so added that as you asked.